executor/src/witgen/jit/block_machine_processor.rs

```diff
@@ -0,0 +1,176 @@
+use std::collections::HashSet;
+
+use powdr_ast::analyzed::AlgebraicReference;
+use powdr_number::FieldElement;
+
+use crate::witgen::{machines::MachineParts, FixedData};
+
+use super::{
+    affine_symbolic_expression::Effect,
+    cell::Cell,
+    witgen_inference::{ReferenceEvaluator, WitgenInference},
+};
+
+struct BlockMachineProcessor<'a, T: FieldElement> {
+    fixed_data: &'a FixedData<'a, T>,
+    machine_parts: MachineParts<'a, T>,
+    block_size: usize,
+    latch_row: usize,
+    inputs: Vec<&'a AlgebraicReference>,
+}
+
+impl<T: FieldElement> BlockMachineProcessor<'_, T> {
+    pub fn generate_code(&self) -> Vec<Effect<T, Cell>> {
+        let known_cells = self
+            .inputs
+            .iter()
+            .map(|input| Cell::from_reference(*input, self.latch_row as i32))
+            .collect::<HashSet<_>>();
+        let ref_eval = ReferenceEvaluatorForFixedData(self.fixed_data);
+        let mut witgen = WitgenInference::new(self.fixed_data, ref_eval, known_cells);
+        let mut complete = HashSet::new();
+        while complete.len() != self.machine_parts.identities.len() * self.block_size {
+            for row in 0..self.block_size {
+                for id in self.machine_parts.identities.iter() {
+                    if !complete.contains(&(id.id(), row))
+                        && witgen.process_identity(id, row as i32)
+                    {
+                        complete.insert((id.id(), row));
+                    }
+                }
+            }
+        }
+        witgen.code()
+    }
+}
+
+// TODO: Copied from witgen_inference.rs, probably FixedData should just implement ReferenceEvaluator.
+struct ReferenceEvaluatorForFixedData<'a, T: FieldElement>(&'a FixedData<'a, T>);
+impl<'a, T: FieldElement> ReferenceEvaluator<T> for ReferenceEvaluatorForFixedData<'a, T> {
+    fn evaluate_fixed(&self, var: &AlgebraicReference, row_offset: i32) -> Option<T> {
+        assert!(var.is_fixed());
+        let values = self.0.fixed_cols[&var.poly_id].values_max_size();
+        let row = (row_offset as usize + var.next as usize) % values.len();
+        Some(values[row])
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::collections::{BTreeMap, HashSet};
+
+    use itertools::Itertools;
+    use powdr_ast::analyzed::{AlgebraicReference, Analyzed};
+    use powdr_number::GoldilocksField;
+
+    use crate::{
+        constant_evaluator,
+        witgen::{
+            jit::{
+                affine_symbolic_expression::{Assertion, Effect, MachineCallArgument},
+                cell::Cell,
+            },
+            machines::MachineParts,
+            FixedData,
+        },
+    };
+
+    use super::BlockMachineProcessor;
+
+    // TODO: Copied
```

Comment:
I'm not sure if it's useful to check for the exact generated code here, it's probably enough to check the 'known cells' after generation and if generation was successful or not.

Reply:
I think it makes sense, because I'll introduce handling the inputs and outputs of the machine call, so it'll be good to have some assertions around that.